src/components/db-explorer.tsx

```diff
@@ -0,0 +1,214 @@
+import { sql } from "@codemirror/lang-sql";
+import { Grid } from "@githubocto/flat-ui";
+import * as duckdb from "@duckdb/duckdb-wasm";
+import CodeMirror from "@uiw/react-codemirror";
+import React, { useEffect, useMemo, useRef, useState } from "react";
+import { useQuery } from "react-query";
+import { useDebounce } from "use-debounce";
+import { useRawDataFile } from "../hooks";
+import { LoadingState } from "./loading-state";
+import { ErrorState } from "./error-state";
+import { Spinner } from "./spinner";
+import Bug from "../bug.svg";
+
+interface Props {
+  sha: string;
+  filename: string;
+  owner: string;
+  name: string;
+}
+
+interface DBExplorerInnerProps {
+  content: string;
+  filename: string;
+  extension: string;
+  sha: string;
+}
+
+const VALID_EXTENSIONS = ["csv", "json"];
+
+function DBExplorerInner(props: DBExplorerInnerProps) {
+  const { content, extension, filename, sha } = props;
+  const connectionRef = useRef<duckdb.AsyncDuckDBConnection | null>(null);
+  const [query, setQuery] = useState("");
+  const [debouncedQuery] = useDebounce(query, 500);
+  const [dbStatus, setDbStatus] = useState<"error" | "idle" | "success">(
+    "idle"
+  );
+
+  const execQuery = async (query: string) => {
+    if (!connectionRef.current) return;
+    const queryRes = await connectionRef.current.query(query);
+    const asArray = queryRes.toArray();
+    return {
+      numRows: queryRes.numRows,
+      numCols: queryRes.numCols,
+      results: asArray.map((row) => {
+        return row.toJSON();
+      }),
+    };
+  };
+
+  const { data, status, error } = useQuery(
+    ["query-results", filename, sha, debouncedQuery],
+    () => execQuery(debouncedQuery),
+    {
+      refetchOnWindowFocus: false,
+      retry: false,
+    }
+  );
+
+  useEffect(() => {
+    const initDuckDb = async () => {
+      const MANUAL_BUNDLES: duckdb.DuckDBBundles = {
+        mvp: {
+          mainModule: "/duckdb/duckdb.wasm",
+          mainWorker: "/duckdb/duckdb-browser-mvp.worker.js",
+        },
+        eh: {
+          mainModule: "/duckdb/duckdb-eh.wasm",
+          mainWorker: "/duckdb/duckdb-browser-eh.worker.js",
+        },
+      };
+      const bundle = await duckdb.selectBundle(MANUAL_BUNDLES);
+      const worker = new Worker(bundle.mainWorker!);
+      const logger = new duckdb.ConsoleLogger();
+      const db = new duckdb.AsyncDuckDB(logger, worker);
+      await db.instantiate(bundle.mainModule, bundle.pthreadWorker);
+
+      const c = await db.connect();
+      connectionRef.current = c;
+
+      try {
+        if (extension === "csv") {
+          await db.registerFileText(`data.csv`, content);
+          await c.insertCSVFromPath(`data.csv`, { name: "data" });
+        } else if (extension === "json") {
+          await db.registerFileText(`data.json`, content);
+          await c.insertJSONFromPath("data.json", { name: "data" });
+        }
```

Comment:
I guess you have plans to change this already, but nevertheless I thought to say that that table names should align with the filename equivalents, for a couple of reasons:
- it's clearer to the user that a query is being made against the current file, especially if the user is having a long session poking at multiple files in a given repo
- it provides an opportunity for subsequent implementations to facilitate joins between files, or even across commits, by somehow creating a DuckDB schema for each commit in a repo.

A rough illustration of how this might look like is demonstrated using the following DuckDB [session](https://shell.duckdb.org/):
```sql
duckdb> CREATE SCHEMA abc1234;
┌┐
└┘
Elapsed: 3 ms

duckdb> CREATE SCHEMA abc1235;
┌┐
└┘
Elapsed: 3 ms

duckdb> CREATE TABLE abc1234."t.csv" AS SELECT 42 AS i, 84 AS j;
┌───────┐
│ Count │
╞═══════╡
│     1 │
└───────┘
Elapsed: 3 ms

duckdb> CREATE TABLE abc1235."t.csv" AS SELECT 42 AS i, 84 AS j;
┌───────┐
│ Count │
╞═══════╡
│     1 │
└───────┘
Elapsed: 3 ms

duckdb> SELECT * FROM abc1234."t.csv" AS a, abc1235."t.csv" AS b where a.i = b.i;
┌────┬────┬────┬────┐
│ i  ┆ j  ┆ i  ┆ j  │
╞════╪════╪════╪════╡
│ 42 ┆ 84 ┆ 42 ┆ 84 │
└────┴────┴────┴────┘
Elapsed: 2 ms


```